### PR TITLE
Days rest: V2

### PR DESCRIPTION
**Pull request by kyleskom**
@nkgilley Here is the updated PR with the fixed UO. If you can train the NN UO model and add it here that would be great. I can't see to get tensorflow to train on my Mac currently.

### CODE REVIEW

File: src/Train-Models/Train_Model_UO.py [hunk 1]
```diff
@@ -17,8 +17,10 @@
 con.close()
 
 OU = data['OU-Cover']
-data.drop(['Score', 'Home-Team-Win', 'TEAM_NAME', 'Date', 'TEAM_NAME.1', 'Date.1', 'OU-Cover'], axis=1, inplace=True)
+total = data['OU']
```
Comment (kyleskom) [inline src/Train-Models/Train_Model_UO.py:20]: This was the fix

File: src/Train-Models/XGBoost_Model_UO.py [hunk 2]
```diff
@@ -12,8 +12,11 @@
 data = pd.read_sql_query(f"select * from \"{dataset}\"", con, index_col="index")
 con.close()
 OU = data['OU-Cover']
-data.drop(['Score', 'Home-Team-Win', 'TEAM_NAME', 'Date', 'TEAM_NAME.1', 'Date.1', 'OU-Cover'], axis=1,
+total = data['OU']
```
Comment (kyleskom) [inline src/Train-Models/XGBoost_Model_UO.py:15]: This was the fix.

File: src/Train-Models/Train_Model_UO.py [hunk 3]
```diff
@@ -17,8 +17,10 @@
...
+data.drop(['Score', 'Home-Team-Win', 'TEAM_NAME', 'Date', 'TEAM_NAME.1', 'Date.1', 'OU-Cover', 'OU'], axis=1, inplace=True)
 
+data['OU'] = np.asarray(total)
 data = data.values
```
Comment (kyleskom) [inline src/Train-Models/Train_Model_UO.py:24]: This was the fix

File: src/Train-Models/XGBoost_Model_UO.py [hunk 4]
```diff
@@ -12,8 +12,11 @@
...
+data.drop(['Score', 'Home-Team-Win', 'TEAM_NAME', 'Date', 'TEAM_NAME.1', 'Date.1', 'OU-Cover', 'OU'], axis=1,
           inplace=True)
+
+data['OU'] = np.asarray(total)
```
Comment (kyleskom) [inline src/Train-Models/XGBoost_Model_UO.py:19]: Fix

**Comment by nkgilley** [comment]
This is awesome!  Glad you were able to get this working.  I just retrained the NN model and opened PR #142

**Comment by kyleskom** [comment]
Gonna give it a few days to see if this looks better. Hard to tell if something still wrong or just actually the predictions for today.
image

**Comment by kyleskom** [comment]
@nkgilley When you trained this new model for the NN did you pull the latest changes from here before training?

**Comment by nkgilley** [comment]
Yea I'm pretty sure that I did.  What looks wrong?

**Comment by kyleskom** [comment]
It just still seems to be returning overs for all games. Ill have to take another deeper look.

**Comment by nkgilley** [comment]
from today's games, it's not all overs:

```
------------Neural Network Model Predictions-----------
Miami Heat vs Memphis Grizzlies (55.2%): OVER 220.5 (52.0%)
Cleveland Cavaliers (55.7%) vs Philadelphia 76ers: UNDER 223 (50.3%)
Chicago Bulls vs Sacramento Kings (57.8%): OVER 236 (49.7%)
Minnesota Timberwolves vs Boston Celtics (61.2%): OVER 233.5 (50.6%)
Houston Rockets vs Los Angeles Lakers (62.8%): OVER 231 (52.8%)
San Antonio Spurs vs Dallas Mavericks (69.8%): OVER 227 (51.7%)
LA Clippers (54.0%) vs Golden State Warriors: OVER 237 (51.6%)
```

**Comment by kyleskom** [comment]
So 2 things are happening
1.) The model and prediction is right but is just so close to 50 or it likes the over slightly more
2.) Its still broken

I think with how close he predictions are to 50% im skeptical.

**Comment by dimmos1** [comment]
can you upload predictions for todays games as mine seems to be playing up

**Comment by Marcfeitosa** [comment]
Hi Kyle,
I've got a quick question. How can i train the model including the 2022 - 2023 season?
Cheers
    Em sexta-feira, 17 de março de 2023 às 09:17:30 BRT, dimmos1 ***@***.***> escreveu:  
 
 


can you upload predictions for todays games as mine seems to be playing up

**Comment by kyleskom** [comment]
@Marcfeitosa @dimmos1 Please use the discussions tab. This is for the open PR.

**Comment by kyleskom** [comment]
Just got back from more travel and going to try to close this this week. Still need to look at the NN model. Also any reason why we don't use the rest days for the ML predictions as well?

**Comment by kyleskom** [comment]
Im having a hard time getting models I trined to run. What version of python, tensorflow-macos and tensorflow-metal are you using?

**Comment by nkgilley** [comment]
Python 3.9.6, tensorflow-macos 2.11.0.  I'm not using tensorflow-metal it doesn't look like, which probably explains why this computer isn't any faster at this than my 10 year old GPU in my old linux box.

```
(.venv) ➜  NBA-Machine-Learning-Sports-Betting git:(days-rest) ✗ python -V
Python 3.9.6
(.venv) ➜  NBA-Machine-Learning-Sports-Betting git:(days-rest) ✗ pip freeze
absl-py==1.4.0
appnope==0.1.3
asttokens==2.2.1
astunparse==1.6.3
backcall==0.2.0
bleach==6.0.0
build==0.10.0
cachetools==5.3.0
certifi==2022.12.7
chardet==4.0.0
charset-normalizer==3.0.1
click==8.1.3
cmake==3.25.0
colorama==0.4.6
decorator==5.1.1
docutils==0.19
et-xmlfile==1.1.0
executing==1.2.0
Flask==2.2.3
flatbuffers==23.1.21
gast==0.4.0
google-auth==2.16.0
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
greenlet==2.0.1
grpcio==1.51.1
h5py==3.8.0
idna==2.10
importlib-metadata==6.0.0
ipython==8.9.0
itsdangerous==2.1.2
jaraco.classes==3.2.3
jedi==0.18.2
Jinja2==3.1.2
joblib==1.2.0
keras==2.11.0
keyring==23.13.1
libclang==15.0.6.1
Markdown==3.4.1
markdown-it-py==2.1.0
MarkupSafe==2.1.2
matplotlib-inline==0.1.6
mdurl==0.1.2
more-itertools==9.0.0
numpy==1.24.2
oauthlib==3.2.2
openpyxl==3.1.1
opt-einsum==3.3.0
packaging==23.0
pandas==1.5.3
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
pkginfo==1.9.6
playwright==1.30.0
prompt-toolkit==3.0.36
protobuf==3.19.6
ptyprocess==0.7.0
pure-eval==0.2.2
pyasn1==0.4.8
pyasn1-modules==0.2.8
pyee==9.0.4
Pygments==2.14.0
pyproject_hooks==1.0.0
python-dateutil==2.8.2
pytz==2022.7.1
readme-renderer==37.3
requests==2.28.2
requests-oauthlib==1.3.1
requests-toolbelt==0.10.1
rfc3986==2.0.0
rich==13.3.1
rsa==4.9
sbrscrape==0.0.6
scikit-learn==1.2.1
scipy==1.10.0
six==1.16.0
stack-data==0.6.2
tensorboard==2.11.2
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow-estimator==2.11.0
tensorflow-macos==2.11.0
termcolor==2.2.0
threadpoolctl==3.1.0
tomli==2.0.1
tqdm==4.64.1
traitlets==5.9.0
twine==4.0.2
typing_extensions==4.4.0
urllib3==1.26.14
wcwidth==0.2.6
webencodings==0.5.1
Werkzeug==2.2.2
wrapt==1.14.1
xgboost==1.7.4
zipp==3.12.0
```

**Comment by kyleskom** [comment]
Im able to train models but when I run them I get an error.

**Comment by nkgilley** [comment]
I can run the models with and without `tensorflow-metal`.  I can't train with it though

**Comment by nkgilley** [comment]
I just got training working with `tensorflow-metal==0.5.0` and `tensorflow-macos==2.9` suggested by this comment: https://developer.apple.com/forums/thread/721619

>I dropped back to the following versions: tensorflow-macos==2.9 and tensorflow-metal==0.5.0. Was using the tensorflow-macos==2.11 and tensorflow-metal==0.7.0 version and just couldn't get things to work. After dropping back I was able to use the GPU and all my validations worked.

I'm able to train and run at much faster speeds now 🚀

**Comment by kyleskom** [comment]
I was able to fix it. It was due to the ML model being out of date. I think the UO on the NN is also fixed

**Comment by kyleskom** [comment]
Wanted to loop back to this quickly 
https://github.com/kyleskom/NBA-Machine-Learning-Sports-Betting/pull/141#issuecomment-1487791147

**Comment by kyleskom** [comment]
Nvm I think its all good now

**Comment by kyleskom** [comment]
Looks like days rest is added and tested for both NN and XGB as well as both ML and UO

**Comment by kyleskom** [comment]
@nkgilley If your still interested in contributing shoot me an email kyleskom@buffalo.edu. Id like to discuss some more things moving forward if your interested.

**Pull request merged**